File: graph_valid_tree/S2.cpp
#include <vector>
#include <iostream>
using namespace std;
//Your submission beats 95.60% Submissions!, by jiadai
class Solution {
public:
    /**
     * @param n an integer
     * @param edges a list of undirected edges
     * @return true if it's a valid tree, or false
     */
    bool validTree(int n, vector<vector<int> >& edges) {
        // Write your code here
        if (n == 0) {
            return true;
        }
        
        if (edges.size() != n - 1) {
            return false;
        }
        
        vector<int> parent(n);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
        
        for (auto edge : edges) {
            int x = edge[0];
            int y = edge[1];
            int xp = findParent(parent, x);
            int yp = findParent(parent, y);
            if (xp == yp) {
                return false;
            }
            
            parent[xp] = yp;
        }
        
        return true;
    }
    
private:
    int findParent(vector<int>& parent, int n) {
        while (parent[n] != n) {
            parent[n] = parent[parent[n]];
            n = parent[n];
        }
        
        return n;
    }
};
int main() {
  Solution sol;
  vector<vector<int, int>> v1{{0, 1}, {0, 2}, {0, 3}, {1, 4}};
  cout << sol.validTree(5, v1) << endl;

  vector<vector<int, int>> v2{{0, 1}, {1, 2}, {2, 3}, {1, 3}, {1, 4}};
  cout << sol.validTree(5, v2) << endl;
  //cout <<"hello"<<endl;
  return 0;
}

File: graph_valid_tree/s2.py
#adapted from jiadai c++, Your submission beats 69.11% Submissions!
#not yet fully understand
class Solution:
    ''' 
     * @param n an integer that is the number of nodes(0 to n-1) in a graph
     * @param edges a list of undirected edges
     * @return true if it's a valid tree, or false
    ''' 
    def validTree(self, n, edges):
        # write your code here

        if n == 0:
            return True

        if len(edges) != n - 1:
            return False

        #vector<int> parent(n);
        parent = [i for i in range(n)]

        for edge in edges:
            x = edge[0]
            y = edge[1]
            xp = self.findParent(parent, x)
            yp = self.findParent(parent, y)
            if xp == yp:
                return False #why? makes sense now since x & y connected, if they has same parent, then there is a cycle
            parent[xp] = yp #  x---y--xp--yp
        return True

    def findParent(self,parent, n):
        while parent[n] != n:
            parent[n] = parent[parent[n]]
            n = parent[n]
        return n
